In [4]:
import pandas as pd
import numpy as np

In [6]:
filename = "DC_Properties.csv"
df = pd.read_csv(filename)

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (18,19,20,21,22,23,24,33,34,35,37,43,45) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


# First Take-Aways

In reading the first output of our dataframe, we see a a warning message of: `C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (18,19,20,21,22,23,24,33,34,35,37,43,45) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)`

Our group thinks it best to research the origin of this warning, and take appropriate action to resolve any issues (such as low memory, or mixed types). Or, perhaps the issue is that numpy wasn't included initially?


In [10]:
df.columns

Index(['Unnamed: 0', 'BATHRM', 'HF_BATHRM', 'HEAT', 'AC', 'NUM_UNITS', 'ROOMS',
       'BEDRM', 'AYB', 'YR_RMDL', 'EYB', 'STORIES', 'SALEDATE', 'PRICE',
       'QUALIFIED', 'SALE_NUM', 'GBA', 'BLDG_NUM', 'STYLE', 'STRUCT', 'GRADE',
       'CNDTN', 'EXTWALL', 'ROOF', 'INTWALL', 'KITCHENS', 'FIREPLACES',
       'USECODE', 'LANDAREA', 'GIS_LAST_MOD_DTTM', 'SOURCE', 'CMPLX_NUM',
       'LIVING_GBA', 'FULLADDRESS', 'CITY', 'STATE', 'ZIPCODE', 'NATIONALGRID',
       'LATITUDE', 'LONGITUDE', 'ASSESSMENT_NBHD', 'ASSESSMENT_SUBNBHD',
       'CENSUS_TRACT', 'CENSUS_BLOCK', 'WARD', 'SQUARE', 'X', 'Y', 'QUADRANT'],
      dtype='object')

In [11]:
df.dtypes

Unnamed: 0              int64
BATHRM                  int64
HF_BATHRM               int64
HEAT                   object
AC                     object
NUM_UNITS             float64
ROOMS                   int64
BEDRM                   int64
AYB                   float64
YR_RMDL               float64
EYB                     int64
STORIES               float64
SALEDATE               object
PRICE                 float64
QUALIFIED              object
SALE_NUM                int64
GBA                   float64
BLDG_NUM                int64
STYLE                  object
STRUCT                 object
GRADE                  object
CNDTN                  object
EXTWALL                object
ROOF                   object
INTWALL                object
KITCHENS              float64
FIREPLACES              int64
USECODE                 int64
LANDAREA                int64
GIS_LAST_MOD_DTTM      object
SOURCE                 object
CMPLX_NUM             float64
LIVING_GBA            float64
FULLADDRES

In [12]:
df.shape

(158957, 49)

In [20]:
df_missing = df.copy()
missing = df_missing.isnull().sum()
missing

Unnamed: 0                 0
BATHRM                     0
HF_BATHRM                  0
HEAT                       0
AC                         0
NUM_UNITS              52261
ROOMS                      0
BEDRM                      0
AYB                      271
YR_RMDL                78029
EYB                        0
STORIES                52305
SALEDATE               26770
PRICE                  60741
QUALIFIED                  0
SALE_NUM                   0
GBA                    52261
BLDG_NUM                   0
STYLE                  52261
STRUCT                 52261
GRADE                  52261
CNDTN                  52261
EXTWALL                52261
ROOF                   52261
INTWALL                52261
KITCHENS               52262
FIREPLACES                 0
USECODE                    0
LANDAREA                   0
GIS_LAST_MOD_DTTM          0
SOURCE                     0
CMPLX_NUM             106696
LIVING_GBA            106696
FULLADDRESS            52917
CITY          

# observations so far

By reviewing the cells above, as well as [the kaggle data page]("https://www.kaggle.com/christophercorrea/dc-residential-properties"), our conclusions is that we are getting some mixed data types, as well as some missing data. We'd like to explore the variation in a selection of columns to start with, such as Heat and AC, as well as attempt to accurately encode date data as dates (not objects) and all string data as strings only (again, not as objects).

# first steps

As a group we've decided to restrict our dataset to the following data columns: BATHRM, HEAT, AC, BEDRMN, PRICE, FULLADDRESS, WARD, QUADRANT

In [23]:
small_df = pd.DataFrame(data = df, columns=['BATHRM', 'HEAT', 'AC', 'BEDRM', 'PRICE', 'FULLADDRESS', 'WARD', 'QUADRANT'])
small_df.head()

BATHRM           HEAT AC  BEDRM      PRICE                   FULLADDRESS  \
0       4      Warm Cool  Y      4  1095000.0          1748 SWANN STREET NW   
1       3      Warm Cool  Y      5        NaN          1746 SWANN STREET NW   
2       3  Hot Water Rad  Y      5  2100000.0          1744 SWANN STREET NW   
3       3  Hot Water Rad  Y      5  1602000.0          1742 SWANN STREET NW   
4       2      Warm Cool  Y      3        NaN  1804 NEW HAMPSHIRE AVENUE NW   

     WARD QUADRANT  
0  Ward 2       NW  
1  Ward 2       NW  
2  Ward 2       NW  
3  Ward 2       NW  
4  Ward 2       NW

In [25]:
small_df.shape

(158957, 8)

# second steps

We want to further condense our dataset, and also make observations about the info we have access to. Therefore, we've decided to remove all rows with NaN in the price column.

In [26]:
no_missing_df = small_df.copy().dropna()
no_missing_df.shape

(57464, 8)

In [29]:
# by counting all the not null entries AND calling shape on our dataset,
# we can confirm that there are no blanks anywhere in this current dataset
no_missing_df.notnull().count()

BATHRM         57464
HEAT           57464
AC             57464
BEDRM          57464
PRICE          57464
FULLADDRESS    57464
WARD           57464
QUADRANT       57464
dtype: int64

In [30]:
no_missing_df.dtypes

BATHRM           int64
HEAT            object
AC              object
BEDRM            int64
PRICE          float64
FULLADDRESS     object
WARD            object
QUADRANT        object
dtype: object

In [46]:
# Just curious to see, what are the unique possible values for AC?
no_missing_df['AC'].unique() # we find that there are properties with an AC value of 0, which doesn't really make sense to us...
no_missing_df.loc[no_missing_df['AC'] == '0'].count() # and we know that there are only 16 such entries, ...
no_missing_df = no_missing_df.drop(no_missing_df[no_missing_df['AC'] == '0'].index) # so we feel unworried about removing these entries to further tighten up our dataset
# take-away here is that we MUST NOT FORGET when dropping rows/entries to use ".index" in the drop parameters
# let's confirm that the result of the above line gives up only 2 unique AC values now
no_missing_df['AC'].unique()

array(['Y', 'N'], dtype=object)

In [49]:
# since Avi wants to live in this type of place (cheap first, good neighborhood 2nd, at least 2 bedroom, 1 bath, under $1500/month.)
# let's first bin the price ranges of places to live...
# before we can effectively bin, we should first describe the dataset to get an idea of the general data distribution


BATHRM         BEDRM         PRICE
count  57447.000000  57447.000000  5.744700e+04
mean       2.201995      3.422024  5.795042e+05
std        1.062376      1.116655  5.979533e+05
min        0.000000      0.000000  1.000000e+00
25%        1.000000      3.000000  2.400000e+05
50%        2.000000      3.000000  4.400000e+05
75%        3.000000      4.000000  7.500000e+05
max       12.000000     20.000000  2.510000e+07

In [51]:
# outstanding questions
# How do we display dates as dates in Jupyter Notebooks?
# How can we filter (condense) the dataset to include sale dates within only the last X years, Y months, Z days, etc.?
# How can we render/display prices in a more human readable output such as $10B, or $200K, etc.?
# As it appears that there was a sale for $1 which seems dubious/suspect, how can we select & show the cheapest properities by sale price? (we are guessing sort & head)
no_missing_df ['PRICE'].min()

1.0

In [52]:
no_missing_df.tail(15)

BATHRM           HEAT AC  BEDRM     PRICE  \
106657       2  Hot Water Rad  Y      3  140496.0   
106662       1      Warm Cool  Y      3  120000.0   
106663       1      Warm Cool  Y      3  109256.0   
106664       1      Warm Cool  Y      3  230000.0   
106666       1      Warm Cool  Y      3  215000.0   
106668       2     Forced Air  Y      3  205000.0   
106672       3     Forced Air  N      6  100000.0   
106673       3     Forced Air  N      6  103000.0   
106682       3     Forced Air  N      8  118000.0   
106683       3     Forced Air  N      6       1.0   
106687       2     Forced Air  N      4   95000.0   
106688       2     Forced Air  Y      4  185000.0   
106692       2     Forced Air  N      5  100000.0   
106694       2     Forced Air  N      4  215000.0   
106695       2     Forced Air  Y      4  140000.0   

                          FULLADDRESS    WARD QUADRANT  
106657         630 MILWAUKEE PLACE SE  Ward 8       SE  
106662           4402 BARNABY ROAD SE  Ward 8       SE  
106663           4404 BARNABY ROAD SE  Ward 8       SE  
106664           4406 BARNABY ROAD SE  Ward 8       SE  
106666           4410 BARNABY ROAD SE  Ward 8       SE  
106668           4414 BARNABY ROAD SE  Ward 8       SE  
106672  4778 SOUTH CAPITOL TERRACE SW  Ward 8       SW  
106673  4774 SOUTH CAPITOL TERRACE SW  Ward 8       SW  
106682  4738 SOUTH CAPITOL TERRACE SW  Ward 8       SW  
106683  4734 SOUTH CAPITOL TERRACE SW  Ward 8       SW  
106687           107 JOLIET STREET SW  Ward 8       SW  
106688           111 JOLIET STREET SW  Ward 8       SW  
106692           127 JOLIET STREET SW  Ward 8       SW  
106694           135 JOLIET STREET SW  Ward 8       SW  
106695           139 JOLIET STREET SW  Ward 8       SW

In [53]:
no_missing_df.head(15)

BATHRM           HEAT AC  BEDRM      PRICE              FULLADDRESS  \
2        3  Hot Water Rad  Y      5  2100000.0     1744 SWANN STREET NW   
3        3  Hot Water Rad  Y      5  1602000.0     1742 SWANN STREET NW   
5        3  Hot Water Rad  Y      5  1950000.0         1709 S STREET NW   
7        3  Hot Water Rad  Y      4  1050000.0     1746 1/2 T STREET NW   
8        3      Warm Cool  Y      3  1430000.0     1727 SWANN STREET NW   
14       3      Warm Cool  Y      3  1325000.0     1721 SWANN STREET NW   
16       3      Warm Cool  Y      3  1240000.0     1725 SWANN STREET NW   
19       3  Hot Water Rad  Y      3   592250.0     1735 SWANN STREET NW   
20       3  Hot Water Rad  Y      5    33232.0         1441 Q STREET NW   
22       1     Forced Air  Y      3   907400.0         1445 Q STREET NW   
23       2     Forced Air  Y      3  1065000.0         1447 Q STREET NW   
24       2  Hot Water Rad  Y      3  1100000.0         1449 Q STREET NW   
25       1  Hot Water Rad  N      4    20000.0         1451 Q STREET NW   
27       3  Hot Water Rad  Y      3   703700.0  1456 CORCORAN STREET NW   
29       3     Forced Air  Y      3  1210000.0  1452 CORCORAN STREET NW   

      WARD QUADRANT  
2   Ward 2       NW  
3   Ward 2       NW  
5   Ward 2       NW  
7   Ward 2       NW  
8   Ward 2       NW  
14  Ward 2       NW  
16  Ward 2       NW  
19  Ward 2       NW  
20  Ward 2       NW  
22  Ward 2       NW  
23  Ward 2       NW  
24  Ward 2       NW  
25  Ward 2       NW  
27  Ward 2       NW  
29  Ward 2       NW